collector: add systemd services status

audebert opened the pull request:
Hi,

I wrote a basic systemd exporter a few weeks ago and saw that #173 requested one, so here is my code. Its main use is to detect failing services.

Feel free to comment :)

brian-brazil commented on collector/systemd.go:
```diff
@@ -0,0 +1,111 @@
+		unitStateCount: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: "systemd",
+				Name:      "unit_state_count",
```
_count should be avoided as a suffix, as it's used by Counters

brian-brazil commented on collector/systemd.go:
```diff
@@ -0,0 +1,111 @@
+	for _, unit := range units {
+		states[unit.ActiveState] += 1
+		if unit.ActiveState == "failed" {
+			unitFailedMap[unit.Name] = 1.0
```
This metrics is redundant with the states.

brian-brazil commented on collector/systemd.go:
```diff
@@ -0,0 +1,111 @@
+	}
+	return &systemdCollector{
+		conn: conn,
+		unitStateCount: prometheus.NewGaugeVec(
```
You should use ConstMetrics like filesystem collector, rather than sharing metrics across scrapes. This handles deleted services and concurrent scrapes better

audebert commented:
Thanks for the review. I'll take some time tonight to fix the code using your comments!

audebert commented:
This new version exports metrics in the following format:

```
# HELP node_systemd_units_active Systemd units in the active state.
# TYPE node_systemd_units_active gauge
node_systemd_units_active{unit="-.mount"} 1
[...]
# HELP node_systemd_units_inactive Systemd units in the inactive state.
# TYPE node_systemd_units_inactive gauge
node_systemd_units_inactive{unit="alsa-state.service"} 1
```

Actually I am not sure if the unit state should be in the metric name or in a "state" label, what do you think?

brian-brazil commented:
A state label would be best.

audebert commented:
Ok :) Here is how the metrics look like now:

```
# HELP node_systemd_unit Systemd unit
# TYPE node_systemd_unit gauge
node_systemd_unit{name="-.mount",state="active"} 1
node_systemd_unit{name="-.slice",state="active"} 1
node_systemd_unit{name="alsa-restore.service",state="inactive"} 1
node_systemd_unit{name="alsa-state.service",state="inactive"} 1
node_systemd_unit{name="apparmor.service",state="inactive"} 1
node_systemd_unit{name="auditd.service",state="inactive"} 1
```

brian-brazil commented:
That looks okay, you'll also want to output a 0 for states it's not in. Dealing with timeseries that appear and disappear is difficult.

audebert commented:
Now:

```
# HELP node_systemd_unit Systemd unit
# TYPE node_systemd_unit gauge
node_systemd_unit{name="-.mount",state="activating"} 0
node_systemd_unit{name="-.mount",state="active"} 1
node_systemd_unit{name="-.mount",state="deactivating"} 0
node_systemd_unit{name="-.mount",state="failed"} 0
node_systemd_unit{name="-.mount",state="inactive"} 0
```

brian-brazil commented on collector/systemd.go:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	unitDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, subsystem, "unit"),
```
`unit_state` would probably be a clearer name, there may be other metrics in the future.

brian-brazil commented on collector/systemd.go:
```diff
@@ -0,0 +1,80 @@
+func NewSystemdCollector() (Collector, error) {
+	var (
+		subsystem  = "systemd"
+		labelNames = []string{"name", "state"}
```
labelNames is only used once, you can remove this variable.

matthiasr commented:
One thing – should this be `systemd_linux.go` so it's not built on other systems automatically? Systemd is very much Linux only at this point.

matthiasr commented:
Can the dbus connection break, and if so, will it be re-established? The systemdCollector will be long-lived and a restart of the exporter should not be required to get it going again in that case.

Also, is it possible to stub out the `ListUnits` call somehow? With that, it would be great to add a test or two. Tying it into the end-to-end test would be a bonus but not sure if that's feasible (the compiled binary would need to be able to be connected to a fake dbus I guess?).

jinty commented:
I'm very happy to see this pull request, I've just been looking for something like this and will give it a test run. A maybe-not-useful suggestion is to measure the result of "is-system-running", a very coarse metric to see if all is well. On a running system:

```
$ sudo systemctl is-system-running && echo "yes: $?" || echo "no: $?"
running
yes: 0
```

and on a system with a failed service:

```
$ sudo systemctl is-system-running && echo "yes: $?" || echo "no: $?"
degraded
no: 1
```

The documentation says; "This returns success (exit code 0) when the system is fully up and running, specifically not in startup, shutdown or maintenance mode, and with no failed services."

audebert commented:
@matthiasr Right now I don't know a reason why the dbus connection could break, nevertheless the code now creates a new connection on each `Update()`.
I renamed the collector to `systemd_linux.go` and added some tests.

audebert commented:
@jinty This is indeed a nice metrics to have. I added a commit for that, can you review it?

juliusv commented on collector/systemd_linux.go:
```diff
@@ -0,0 +1,114 @@
+	)
+	systemRunningDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, subsystem, "system_running"),
+		"Whether the system is operational (see `systemctl is-system-running')",
```
Is the backtick+single-quote quoting intentional here?

juliusv commented on collector/systemd_linux.go:
```diff
@@ -0,0 +1,114 @@
+		isSystemRunning = 1.0
+	}
+	ch <- prometheus.MustNewConstMetric(c.systemRunningDesc, prometheus.GaugeValue,
+		isSystemRunning)
```
No need for this kind of line break here, as in Go there's usually no strict limit. It's more common to either put all args on a single line or put them each on their own line if the list gets very long. Single line should be fine here.

audebert commented:
@juliusv I have updated the commits taking into account your review, thanks :)

juliusv commented:
@audebert Cool :) Oh forgot, also add it to the collector list in the README.md please: https://github.com/prometheus/node_exporter#disabled-by-default.

audebert commented:
@juliusv done!

juliusv commented:
:+1: Thanks :)

@matthiasr did you want more tests here or good for now? (obviously, more tests are always good)

matthiasr commented:
:+1: LGTM
